Fix error with 'mysql_native_password' option when attempting to bring up a mysql container

Trentr9 opened the pull request:
## Description
When MYSQL 8.4 was released it it no longer supports the `default-authentication-plugin=mysql_native_password` configuration, instead you need to specify `mysql_native_password=on` from MYSQL version 8.4 onwards.

This was attempted to be fixed in pull request #3521 but this broke it completely. There are a number of issues on this pull request;
- MYSQL_VERSION won't just be a number, it can be 'latest' (this is the default open in Laradock)
- To compare floats (in this case, version strings), it requires some funkyness which has been implemented, because the previous check didn't do that, it caused the script to also go to the true block no matter what the version of MySQL was
- Missing semi-colon on line 19 that caused the rest of the bash script to be echo'd to the file, so even if the if statement did work as expected the config file would still be wrong

This pull request implements logic into the Dockerfile for the MySQL container to check what version of MySQL is running and fixes this issue.

## Motivation and Context
This fixes the problem of the MySQL container not getting stood up no matter what version of MySQL you try to use.

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue).
- [] New feature (non-breaking change which adds functionality).
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected).

## Definition of Done Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [X] I've read the [Contribution Guide](http://laradock.io/contributing).
- [] I've updated the **documentation**. (refer to [this](http://laradock.io/contributing/#update-the-documentation-site) for how to do so).
- [X] I enjoyed my time contributing and making developer's life easier :)

Trentr9 committed:
fix error when attempting to bring up a mysql container.

grooverdan commented:
A simpler way, as part of my.cnf:
```
loose-default-authentication-plugin=mysql_native_password
```
With loose the invalid option is ignored by the server.

Or you could put
```
[mysql-8.4]
mysql_native_password=on
```

Trentr9 commented:
I did not know that about that configuration options.

Agreed that it might be the simpler solution, not sure if that will change in the future if they change the config....

Will leave this open now for the maintainers to see what direction that'd like to go to solve this issue.

Cheers!

rc1021 commented:
You are right, checking the MySQL version should not rely on the MYSQL_VERSION environment variable to determine whether to enable mysql_native_password. It must use the installed system version to make the determination.

```
RUN MYSQL_VERSION=$(mysqld --version | grep -o -E "[0-9]+\.[0-9]+\.[0-9]+") && \
    if [ "$(printf '%s\n' "8.4.0" "$MYSQL_VERSION" | sort -V | head -n1)" = "8.4.0" ]; then \
        echo "mysql_native_password=on" >> /etc/mysql/conf.d/my.cnf; \
    else \
        echo "default-authentication-plugin=mysql_native_password" >> /etc/mysql/conf.d/my.cnf; \
    fi
```

rc1021 commented:
@Trentr9 You are right, checking the MySQL version should not rely on the MYSQL_VERSION environment variable to determine whether to enable mysql_native_password. It must use the installed system version to make the determination.

rc1021 reviewed (changes requested):
```
RUN MYSQL_VERSION=$(mysqld --version | grep -o -E "[0-9]+\.[0-9]+\.[0-9]+") && \
    if [ "$(printf '%s\n' "8.4.0" "$MYSQL_VERSION" | sort -V | head -n1)" = "8.4.0" ]; then \
        echo "mysql_native_password=on" >> /etc/mysql/conf.d/my.cnf; \
    else \
        echo "default-authentication-plugin=mysql_native_password" >> /etc/mysql/conf.d/my.cnf; \
    fi
```

patrick-cullen commented:
> ```
> RUN MYSQL_VERSION=$(mysqld --version | grep -o -E "[0-9]+\.[0-9]+\.[0-9]+") && \
>     if [ "$(printf '%s\n' "8.4.0" "$MYSQL_VERSION" | sort -V | head -n1)" = "8.4.0" ]; then \
>         echo "mysql_native_password=on" >> /etc/mysql/conf.d/my.cnf; \
>     else \
>         echo "default-authentication-plugin=mysql_native_password" >> /etc/mysql/conf.d/my.cnf; \
>     fi
> ```

For others that ran into this problem of not being able to build the mysql container (getting this error: `[ERROR] unknown variable 'mysql_native_password=on else echo default-authentication-plugin=mysql_native_password')`  - replacing the following block in `laradock/mysql/Dockerfile` with @rc1021 's solution above and rebuilding my mysql container with `docker compose build --no-cache mysql` worked for me (MYSQL_VERSION=5.7, in my case): 

```
RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \
  echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf \
else \
  echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf \
;fi
```

Trentr9 committed:
update mysql version check

updated mysql version check to check the installed version instead of the environment variable

Trentr9 commented:
@rc1021 Nice pickup -- I though I was checking the installed MYSQL version but I am not. Must've been up too late doing this!

You're requested change looks great, except when they release 8.4.1 or above it will break again so I have modified my original code to looks at the installed MYSQL version as it should've in the first place, but still does the magic to see if the version installed is newer than 8.3 :)

Chung-NB commented:
Someone can please tell me how to fix this, step by step. I've read everything but still cannot fix this error.

sang-hv commented:
> Someone can please tell me how to fix this, step by step. I've read everything but still cannot fix this error.

same, on mac m1 :( @grooverdan can you help us

grooverdan commented:
There's a merged fix so not sure why this is still open. If still having trouble, use MariaDB instead.

Trentr9 commented:
@grooverdan that merge actually broke it...

Trentr9 commented:
@Chung-NB @sang-hv 

You could run
`git clone https://github.com/Trentr9/laradock.git`

That will clone the repo from me that has the fix that I'm waiting for to be merged in

grooverdan commented:
I recommended a very simple configuration option only solution. Not sure why everyone's feeling compelled to write version checks script when the server can already to version checking.

sang-hv commented:
> I recommended a very simple configuration option only solution. Not sure why everyone's feeling compelled to write version checks script when the server can already to version checking.
@grooverdan I also don't know why, when I build on mac mini M1 - it works but on mac pro m1 it fails
`unknown variable 'mysql_native_password=on else echo default-authentication-plugin=mysql_native_password'`

sang-hv commented:
> @Chung-NB @sang-hv
> 
> You could run `git clone https://github.com/Trentr9/laradock.git`
> 
> That will clone the repo from me that has the fix that I'm waiting for to be merged in

thank you!

grooverdan commented:
> > I recommended a very simple configuration option only solution. Not sure why everyone's feeling compelled to write version checks script when the server can already to version checking.
> > @grooverdan I also don't know why, when I build on mac mini M1 - it works but on mac pro m1 it fails

reading error message is a good way to understand.

> > `unknown variable 'mysql_native_password=on else echo default-authentication-plugin=mysql_native_password'`

That's not what I said. For the avoidance of doubt - #3530.

SPC2356789 commented:
我根據樓主的解法，已經成功運行，最重要的是修改DockerFile

Kenini1805 commented:
@sang-hv @Chung-NB 
Try 
```
docker volume rm $(docker volume ls -q)
```
After that run 
```
docker-compose up -d mysql
```
From VietNam fucking with love

dmcooder commented:
Hello,

In mysql Dockerfile I changed from this:

`
RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \
  echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf \
else \
  echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf \
;fi
`

to this

`
RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \
  echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf; \
else \
  echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf; \
fi
`



It works for me

rivara commented:
> Hello,
> 
> In mysql Dockerfile I changed from this:
> 
> `RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \ echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf \ else \ echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf \ ;fi`
> 
> to this
> 
> `RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \ echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf; \ else \ echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf; \ fi`
> 
> It works for me

is the same!!

Lymhit commented:
> > Hello,
> > In mysql Dockerfile I changed from this:
> > `RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \ echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf \ else \ echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf \ ;fi`
> > to this
> > `RUN if [ ${MYSQL_VERSION} > '8.4.0-0.000' ]; then \ echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf; \ else \ echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf; \ fi`
> > It works for me
> 
> is the same!!

Missing ";" after my.cnf

TFnight commented:
```
RUN if [[ ${MYSQL_VERSION} > '8.4.0-0.000' && ${MYSQL_VERSION} < '9.0.0-0.000' ]]; then \ 
  echo 'mysql_native_password=on' >> /etc/mysql/conf.d/my.cnf; \ 
elif [ ${MYSQL_VERSION} > '9.0.0-0.000' ]; then \ 
  : \
else \ 
  echo 'default-authentication-plugin=mysql_native_password' >> /etc/mysql/conf.d/my.cnf; \ 
fi
```

it works for me.  could give it a try?

If you use MYSQL_VERSION=latest, you'll get the newest MySQL 9.0 after July 3, 2024. In this version, the configuration mysql_native_password=on will cause the MySQL container to crash.

maybe due to [this](https://dev.mysql.com/doc/relnotes/mysql/9.0/en/news-9-0-0.html#:~:text=Microsoft%20Windows%3A%20On%20Windows%2C%20MySQL%20Configurator%20now%20detects,authentication%20plugin%20before%20performing%20a%20MySQL%20Server%20upgrade.%20(WL%20%2316139))?

parallels999 commented:
try #3540

bantawa04 commented:
> try #3540

doesn't work

parallels999 commented:
Did you change `.env`?
```diff
-MYSQL_VERSION=latest
+# 5.7, 8.0, 8.4, 9.0
+MYSQL_VERSION=8.4
```
